lib/Scheduling/CPSATSchedulers.cpp

```diff
@@ -0,0 +1,138 @@
+
+#include "circt/Scheduling/Algorithms.h"
+
+#include "mlir/IR/Operation.h"
+
+#include "ortools/sat/cp_model.h"
+
+#include "llvm/Support/Debug.h"
+#include "llvm/Support/Format.h"
+
+#define DEBUG_TYPE "cpsat-schedulers"
+
+using namespace circt;
+using namespace circt::scheduling;
+using namespace operations_research;
+using namespace operations_research::sat;
+
+using llvm::dbgs;
+using llvm::format;
+
+LogicalResult scheduling::scheduleCPSAT(SharedOperatorsProblem &prob,
+                                        Operation *lastOp) {
+  Operation *containingOp = prob.getContainingOp();
+  if (!prob.hasOperation(lastOp))
+    return containingOp->emitError("problem does not include last operation");
+
+  CpModelBuilder cpModel;
+  auto &tasks = prob.getOperations();
+  auto &resources = prob.getOperatorTypes();
+
+  DenseMap<Operation *, IntVar> taskStarts;
+  DenseMap<Operation *, IntVar> taskEnds;
+  DenseMap<Problem::OperatorType, llvm::SmallVector<IntervalVar, 4>>
+      resourcesToTaskIntervals;
+
+  // first get horizon
+  unsigned horizon = 0;
+  for (auto *task : tasks) {
+    unsigned duration = *prob.getLatency(*prob.getLinkedOperatorType(task));
+    horizon += duration;
+  }
+
+  // build vars
+  for (auto item : llvm::enumerate(tasks)) {
+    auto i = item.index();
+    auto *task = item.value();
+    IntVar startVar = cpModel.NewIntVar(Domain(0, horizon))
+                          .WithName((Twine("start_of_task_") + Twine(i)).str());
+    IntVar endVar = cpModel.NewIntVar(Domain(0, horizon))
+                        .WithName((Twine("end_of_task_") + Twine(i)).str());
+    taskStarts[task] = startVar;
+    taskEnds[task] = endVar;
+    auto resource = prob.getLinkedOperatorType(task);
+    if (resource.hasValue()) {
+      unsigned duration = *prob.getLatency(*resource);
+      IntervalVar taskInterval =
+          cpModel.NewIntervalVar(startVar, duration, endVar)
+              .WithName((Twine("task_interval_") + Twine(i)).str());
+
+      if (!resourcesToTaskIntervals.count(resource.value()))
+        resourcesToTaskIntervals[resource.value()] = {};
+      resourcesToTaskIntervals[resource.value()].emplace_back(taskInterval);
+    }
+  }
+
+  IntVar makeSpan = cpModel.NewIntVar(Domain(0, horizon)).WithName("makespan");
+  for (Operation *task : tasks) {
+    cpModel.AddLessOrEqual(taskEnds[task], makeSpan);
+    for (auto dep : prob.getDependences(task)) {
+      Operation *src = dep.getSource();
+      Operation *dst = dep.getDestination();
+      if (src == dst)
+        return containingOp->emitError() << "dependence cycle detected";
+      cpModel.AddLessOrEqual(taskEnds[src], taskStarts[dst]);
+    }
+  }
+
+  for (auto resourceToTaskIntervals : resourcesToTaskIntervals) {
+    // https://or.stackexchange.com/a/3363
+    const Problem::OperatorType &resource = resourceToTaskIntervals.getFirst();
+    auto capacity = prob.getLimit(resource);
+    if (!capacity)
+      continue;
```

Comment:
Efficiency nit: you could also not build and collect the intervals for unlimited resources in the first place.

Reply:
So there's an issue here that revolves around `"barrier"(...) { opr = "_0", problemStartTime = 10 }` and `{ name = "_0", latency = 0 }` (i.e. no `limit = ...`). It's the issue that was tripping me [here](https://github.com/llvm/circt/pull/3350#discussion_r973097087). If I do

```cpp
  for (auto item : llvm::enumerate(tasks)) {
    ...
    auto resource = prob.getLinkedOperatorType(task);
    if (resource.hasValue()) {
      auto capacity = prob.getLimit(*resource);
      if (!capacity) {
        continue;
      }
      ...
      resourcesToTaskIntervals[resource.value()].emplace_back(taskInterval);
    }
  }
```

then neither the barriers nor the `return` op will be considered in the problem. The `return` will be handled by 

```cpp
cpModel.AddLessOrEqual(taskEnds[lastOp], makeSpan);
```

but the `"barrier"` op isn't *naturally* handled (I could of course include some string checking to catch `"barrier"`).

Reply:
Ah ok, seems you have to at least _construct_ the interval variable in order to tie a task's start, end, and duration together (but I still think you don't have to store it in the map, as it only used again if the resource is limited).

Reply:
> but the "barrier" op isn't naturally handled (I could of course include some string checking to catch "barrier").

No, don't do that, the (dummy!) barrier instruction from the test case shouldn't receive any special handling or mention in your code here.

Reply:
> Ah ok, seems you have to at least _construct_ the interval variable in order to tie a task's start, end, and duration together (but I still think you don't have to store it in the map, as it only used again if the resource is limited).

This did the trick.